src/providers.rs

```diff
@@ -427,10 +426,13 @@ pub fn do_register_provider(caller: Principal, args: RegisterProviderArgs) -> u6
         m.borrow_mut().set(metadata).unwrap();
         id
     });
-    do_deauthorize(caller, Auth::RegisterProvider);
     log!(INFO, "[{}] Registering provider: {:?}", caller, provider_id);
-    PROVIDERS.with(|providers| {
-        providers.borrow_mut().insert(
+    PROVIDERS.with_borrow_mut(|providers| {
+        assert!(
+            providers.len() <= MAXIMUM_PROVIDER_COUNT,
+            "Too many providers"
```

Comment:
I'm a bit confused by this new limit, which IIUC is a _global_ limit. Because it seems to me that a principal could then register 10_000 providers and then no one else would be able to register a principal (DoS). Am I missing something here?

It's currently not immediately clear to me why this is part of this PR (I don't see the link with the mutability risk). Unless there is an immediate problem with the current approach, I would be in favour of splitting this part to another PR, which would be less time critical.

Reply:
I agree with Grégory that it's not clear why this change is part of this PR.
I think it's a non-trivial issue: It's not great if the number is unbounded. On the other hand, imposing a hard limit introduces the risk that Grégory mentioned.

Reply:
The connection is that because it's now necessary to register new providers in place of updating them, it no longer makes sense (IMO) to require an NNS proposal for each new provider. The global limit is a backup measure in case a `Manage` principal is compromised. A DoS would likely be unavoidable if this principal is compromised in general.